test/common/common/inline_map_speed_test.cc

```diff
@@ -8,17 +8,182 @@ namespace Envoy {
 namespace {
 
 // NOLINTNEXTLINE(readability-identifier-naming)
-static void BM_InlineMapFind(benchmark::State& state) {
-  size_t map_type = state.range(0);
-  size_t used_inline_handles = state.range(1);
+static void BM_InlineMapConstructAndDestruct(benchmark::State& state) {
+  const size_t map_type = state.range(0);
+
+  InlineMapDescriptor<std::string> descriptor_200;
+  // Create 200 inline keys.
+  for (size_t i = 0; i < 200; ++i) {
+    descriptor_200.addInlineKey(absl::StrCat("key_", i));
+  }
+  InlineMapDescriptor<std::string> descriptor_0;
+
+  descriptor_200.finalize();
+  descriptor_0.finalize();
+
+  const auto create_normal_map = []() {
+    absl::flat_hash_map<std::string, std::string> normal_map;
+    return normal_map.size();
+  };
+
+  const auto create_inline_map_200 = [&descriptor_200]() {
+    InlineMap<std::string, std::string> inline_map(descriptor_200);
+    return inline_map.size();
+  };
+
+  const auto create_inline_map_0 = [&descriptor_0]() {
+    InlineMap<std::string, std::string> inline_map(descriptor_0);
+    return inline_map.size();
+  };
+
+  if (map_type == 0) {
+    // Normal map.
+    for (auto _ : state) { // NOLINT
+      benchmark::DoNotOptimize(create_normal_map());
+    }
+  } else if (map_type == 1) {
+    // Inline map without any inline keys.
+    for (auto _ : state) { // NOLINT
+      benchmark::DoNotOptimize(create_inline_map_0());
```

Comment:
i would not use lambdas in this perf benchmark if you can avoid it -- the thing you are doing is so quick that the lambda overhead may distort things.

I don't think this affects the vaiidity of the test, but can you just hand-inline the body of the lambda? 

One thing you can do in lieu of DoNotOptimize  (which requires an expression) here is just sum the size over the whole exercise and then RELEASE_ASSERT that it's 0.

Reply:
got it, will try that and update above benchmark results.

I think another things need to be resolved is the alignment of the values. I just read source code of absl::flat_hash_map in more detail, Now I know why it place flags on the front of the values: for the alignment.

Reply:
Yeah a huge difference is that absl::flat_hash_map is the evolution over decades of a hash table whose usage consumes a meaningful fraction of the world's compute. So it's worth it to hyper-optimize it at the expense of readability and tricky alignments.

I think erring on the side of simplicity is better in this case where so far we have not found profitable usage.

Reply:
Just on the alignment point, I am worring about the correctness/potential bug rather than performance. Because we ignore the align requirement of type `Value` completely now.
Although from the test results, it should be ok on arm/x64 cpu.

Reply:
```
--------------------------------------------------------------------------------------------
Benchmark                                                  Time             CPU   Iterations
--------------------------------------------------------------------------------------------
BM_InlineMapConstructAndDestruct/0                     0.000 ns        0.000 ns   1000000000
BM_InlineMapConstructAndDestruct/1                      8.10 ns         8.10 ns     82709847
BM_InlineMapConstructAndDestruct/2                      8.13 ns         8.13 ns     85908102
BM_InlineMapConstructAndDestructWithSingleEntry/0        131 ns          131 ns      5370223
BM_InlineMapConstructAndDestructWithSingleEntry/1        201 ns          201 ns      3579834
BM_InlineMapConstructAndDestructWithSingleEntry/2        315 ns          315 ns      2169680
BM_InlineMapSet/0/0                                    12547 ns        12547 ns        56556
BM_InlineMapSet/1/0                                    14863 ns        14863 ns        46503
BM_InlineMapSet/2/0                                     7195 ns         7195 ns        97024
BM_InlineMapSet/2/1                                     2439 ns         2439 ns       286474
BM_InlineMapGet/0/0                                     4711 ns         4711 ns       150824
BM_InlineMapGet/1/0                                     5003 ns         5003 ns       138826
BM_InlineMapGet/2/0                                     6523 ns         6523 ns       107600
BM_InlineMapGet/2/1                                    0.000 ns        0.000 ns   1000000000
```

Reply:
> The simplest thing is to keep the allocations separate as that's obviously correct, unless a benchmark calls out a benefit to combing them. And then justify (in code comments) doing the alignment work with an a/b test.

Nope, separate allocations connot resolve this problem. Because we can only allocate memory by primitive type (uint8, uint16, etc) and then construct the `Value` by placement new. So the memory alignment requirment of the primitive type will be used when we allocate the memory rather the the requirement of `Value` type.

It possbile the the `Value` type require a 16/32 bytes alignment requirement.

```
// Just for example.
struct alignas(16) ValueStruct {};
```

Or we can use a container (vector or fixed array) to manage these values, but this way has other problems according our previous discussion. (Although I forget what it is.)

Reply:
> Which reminds me, have you patched this update into your larger change that actualy uses it for filter state to see if there's a win for Envoy?

Nope, orz, I basically have abandoned the filter state for now. I initially want to try the per_filter_config in my own folk after this PR is landed.

Reply:
What about something like:
```
   Value* value_;
  ...
   value_ = reintepret_cast<Value*>(malloc(sizeof(Value[size])));
```
Per https://man7.org/linux/man-pages/man3/malloc.3.html , malloc returns a pointer to the allocated memory, which is suitably aligned for any type that fits into the requested size or less.

I guess you could also use new char[sizeof(Value[size])];

Reply:
Now `new uint8_t[]` is used. If the `malloc` has ensured the alignment, then I think we needn't change anything for now.

Although I accutally doubt that (or I cannot convince my self why flat_hash_map need so complex alignment). Here is a similiar discussion in the https://stackoverflow.com/questions/38088732/explanation-to-aligned-malloc-implementation.

I think there are two possible solutions for now:
1. handle the alignment manually, like the flat_hash_map. Do the same thing should be OK and safe. But this will penalize the readibility.
2. use container like fixed_array to manage the flags and values. It may penalize performance (create/destruct), but will be easy to maintain.

Reply:
I am not sure if `new uint8_t[]` is guaranteed to allow for 8-byte aligned objects. Is there doc that says that explicitly? It might be but it's not obvious. I think basic implementations will use malloc and inherit its properties but idk if that is guaranteed. It's probably fine, but suppose someone overrides operator new on this type. I have actually done similar things for char-buffers in the past to avoid wasting bytes.

Another reason to use separate allocations is to help provide tools like asan more information to track writing/reading past the end of the array.

Reply:
> I guess you could also use new char[sizeof(Value[size])];

If unit_8[] cannot ensure that memory that aligned with 8bytes are allocated, I think char[] cannot also.

Reply:
I think I am convinced to do separate allocations. But still don't know how to handle the alignment.